File: cli/lib/remotion.js
const { getDb, saveDb } = require('./store');

const REMOTION_COMPOSITIONS = [
  {
    id: "KineticTypography",
    alias: "kinetic",
    displayName: "Teks Animasi Bergaya Kinetik & Tipografi Dinamis",
    description: "Animasi teks 3D kinetik dengan efek staggered words, spring bounce, dan glowing neon",
    targetTrackType: "TEXT",
    defaultDurationMs: 4000,
    defaultProps: {
      title: "FLOWMONKEY PRO",
      subtitle: "PRO CLOUD RENDERING ENGINE",
      themeColor: "#8B5CF6",
      accentColor: "#06B6D4"
    }
  },
  {
    id: "GlitchWaveTransition",
    alias: "glitch",
    displayName: "Transisi Glitch Digital & Distorsi Gelombang",
    description: "Distorsi sinusoidal, slice horizontal block displacement, RGB chromatic split & scanlines",
    targetTrackType: "VIDEO",
    defaultDurationMs: 2500,
    defaultProps: {
      glitchIntensity: 0.85,
      waveSpeed: 1.5,
      chromaticShiftPx: 18,
      scanlines: true,
      glitchColor: "#EC4899"
    }
  },
  {
    id: "HudInfographics",
    alias: "hud",
    displayName: "Komponen Infografis, Chart Bergerak, & Elemen HUD",
    description: "HUD radar putar, grafik batang analitik dinamis, crosshair targeting & telemetry readout",
    targetTrackType: "TEXT",
    defaultDurationMs: 5000,
    defaultProps: {
      hudTitle: "TACTICAL TELEMETRY HUD",
      statValue: 98.4,
      dataPoints: [25, 45, 60, 80, 75, 98],
      radarSpeed: 1.2,
      hudColor: "#00E5FF"
    }
  },
  {
    id: "ParticleGlowVfx",
    alias: "particle",
    displayName: "Partikel Melayang & Efek Cahaya (Glow & Particle Systems)",
    description: "Simulasi partikel dinamis dengan pergerakan acak, neon glow halo, dan ambient pulsing vignette",
    targetTrackType: "VIDEO",
    defaultDurationMs: 6000,
    defaultProps: {
      particleCount: 80,
      particleSpeed: 1.4,
      glowIntensity: 1.8,
      particleColor: "#F59E0B",
      particleType: "cyber_embers"
    }
  }
];

function listRemotionVfx(jsonMode = false) {
  if (jsonMode) {
    return JSON.stringify({ compositions: REMOTION_COMPOSITIONS }, null, 2);
  }

  let output = `\n✨ VERCEL + REMOTION CLOUD VFX COMPOSITIONS (SERVERLESS GPU CLUSTER)\n========================================================================\n`;
  REMOTION_COMPOSITIONS.forEach((comp, idx) => {
    output += `${idx + 1}. [${comp.alias} / ${comp.id}]\n`;
    output += `   Nama        : ${comp.displayName}\n`;
    output += `   Tipe Track  : ${comp.targetTrackType} Track | Durasi Default: ${comp.defaultDurationMs / 1000}s\n`;
    output += `   Deskripsi   : ${comp.description}\n\n`;
  });
  output += `========================================================================\n`;
  return output;
}

function renderRemotionCloudVfx({
  vfxType = "kinetic",
  title = "FLOWMONKEY CLOUD VFX",
  subtitle = "RENDERED WITH REMOTION ON VERCEL",
  themeColor = "#8B5CF6",
  durationSec = null,
  trackIndex = null,
  jsonMode = false
}) {
  const db = getDb();
  const activeProjId = db.activeProjectId || (db.projects[0] ? db.projects[0].id : 1);

  const matchedComp = REMOTION_COMPOSITIONS.find(
    c => c.id.toLowerCase() === vfxType.toLowerCase() || c.alias.toLowerCase() === vfxType.toLowerCase()
  ) || REMOTION_COMPOSITIONS[0];

  const durationMs = durationSec ? (parseFloat(durationSec) * 1000) : matchedComp.defaultDurationMs;

  // Find or create target track
  let track = db.tracks.find(t => t.projectId === activeProjId && t.type === matchedComp.targetTrackType);
  if (!track) {
    const newTrackId = db.tracks.length > 0 ? Math.max(...db.tracks.map(t => t.id)) + 1 : 1;
    track = {
      id: newTrackId,
      projectId: activeProjId,
      trackIndex: db.tracks.filter(t => t.projectId === activeProjId).length,
      type: matchedComp.targetTrackType,
      label: `Remotion Cloud VFX (${matchedComp.targetTrackType})`
    };
    db.tracks.push(track);
  }

  // Calculate start time based on existing clips in track
  const trackClips = db.clips.filter(c => c.projectId === activeProjId && c.trackId === track.id);
  const startTimeMs = trackClips.length > 0 ? Math.max(...trackClips.map(c => c.endTimeMs)) : 0;
  const endTimeMs = startTimeMs + durationMs;

  const jobId = `remotion_${matchedComp.alias}_${Date.now()}`;
  const cloudCdnUrl = `https://storage.googleapis.com/flowmonkey-vfx-cloud/${matchedComp.id.toLowerCase()}_${jobId}.mp4`;

  const newClipId = db.clips.length > 0 ? Math.max(...db.clips.map(c => c.id)) + 1 : 1;
  const newClip = {
    id: newClipId,
    projectId: activeProjId,
    trackId: track.id,
    title: `[Remotion Cloud] ${matchedComp.id}: ${title.substring(0, 20)}`,
    uri: cloudCdnUrl,
    startTimeMs,
    endTimeMs,
    durationMs,
    filterName: "Remotion Cloud HDR",
    effectName: matchedComp.id,
    speedMultiplier: 1.0,
    speedCurve: "Standard",
    hasKeyframe: false,
    keyframes: [],
    transitionType: "None",
    volume: 1.0,
    isMuted: false,
    noiseReduction: false,
    vocalEnhance: false,
    proxyUri: `proxy_${jobId}.mp4`,
    proxyStatus: "READY",
    cloudMetadata: {
      renderer: "Vercel Serverless + Remotion GPU Cluster",
      compositionId: matchedComp.id,
      inputProps: {
        title,
        subtitle,
        themeColor
      },
      renderedAt: new Date().toISOString()
    }
  };

  db.clips.push(newClip);
  saveDb(db);

  if (jsonMode) {
    return JSON.stringify({
      success: true,
      jobId,
      compositionId: matchedComp.id,
      clip: newClip,
      streamUrl: cloudCdnUrl
    }, null, 2);
  }

  return `
🚀 VERCEL + REMOTION CLOUD VFX SELESAI DI-RENDER!
========================================================================
Komposisi   : ${matchedComp.displayName} (${matchedComp.id})
Job ID      : ${jobId}
Judul/Teks  : "${title}"
Warna Tema  : ${themeColor}
Durasi      : ${(durationMs / 1000).toFixed(1)}s (Track ${track.type} #${track.id})
Timeline    : Ditambahkan pada rentang ${(startTimeMs / 1000).toFixed(1)}s - ${(endTimeMs / 1000).toFixed(1)}s
Cloud URL   : ${cloudCdnUrl}
========================================================================
`;
}

function previewRemotionCloudVfx(vfxType = "kinetic", frame = 30, jsonMode = false) {
  const matchedComp = REMOTION_COMPOSITIONS.find(
    c => c.id.toLowerCase() === vfxType.toLowerCase() || c.alias.toLowerCase() === vfxType.toLowerCase()
  ) || REMOTION_COMPOSITIONS[0];

  const previewThumbnail = `https://storage.googleapis.com/flowmonkey-vfx-cloud/previews/${matchedComp.id}_f${frame}.jpg`;

  if (jsonMode) {
    return JSON.stringify({
      success: true,
      compositionId: matchedComp.id,
      previewFrame: parseInt(frame),
      previewThumbnail,
      timestamp: new Date().toISOString()
    }, null, 2);
  }

  return `🖼️ Keyframe Preview [Frame #${frame}] untuk '${matchedComp.id}':\n   URL: ${previewThumbnail}`;
}

module.exports = {
  REMOTION_COMPOSITIONS,
  listRemotionVfx,
  renderRemotionCloudVfx,
  previewRemotionCloudVfx
};

File: cli/lib/remotion.py
import json
import time
from lib.store import get_db, save_db

REMOTION_COMPOSITIONS = [
    {
        "id": "KineticTypography",
        "alias": "kinetic",
        "displayName": "Teks Animasi Bergaya Kinetik & Tipografi Dinamis",
        "description": "Animasi teks 3D kinetik dengan efek staggered words, spring bounce, dan glowing neon",
        "targetTrackType": "TEXT",
        "defaultDurationMs": 4000,
        "defaultProps": {
            "title": "FLOWMONKEY PRO",
            "subtitle": "PRO CLOUD RENDERING ENGINE",
            "themeColor": "#8B5CF6",
            "accentColor": "#06B6D4"
        }
    },
    {
        "id": "GlitchWaveTransition",
        "alias": "glitch",
        "displayName": "Transisi Glitch Digital & Distorsi Gelombang",
        "description": "Distorsi sinusoidal, slice horizontal block displacement, RGB chromatic split & scanlines",
        "targetTrackType": "VIDEO",
        "defaultDurationMs": 2500,
        "defaultProps": {
            "glitchIntensity": 0.85,
            "waveSpeed": 1.5,
            "chromaticShiftPx": 18,
            "scanlines": True,
            "glitchColor": "#EC4899"
        }
    },
    {
        "id": "HudInfographics",
        "alias": "hud",
        "displayName": "Komponen Infografis, Chart Bergerak, & Elemen HUD",
        "description": "HUD radar putar, grafik batang analitik dinamis, crosshair targeting & telemetry readout",
        "targetTrackType": "TEXT",
        "defaultDurationMs": 5000,
        "defaultProps": {
            "hudTitle": "TACTICAL TELEMETRY HUD",
            "statValue": 98.4,
            "dataPoints": [25, 45, 60, 80, 75, 98],
            "radarSpeed": 1.2,
            "hudColor": "#00E5FF"
        }
    },
    {
        "id": "ParticleGlowVfx",
        "alias": "particle",
        "displayName": "Partikel Melayang & Efek Cahaya (Glow & Particle Systems)",
        "description": "Simulasi partikel dinamis dengan pergerakan acak, neon glow halo, dan ambient pulsing vignette",
        "targetTrackType": "VIDEO",
        "defaultDurationMs": 6000,
        "defaultProps": {
            "particleCount": 80,
            "particleSpeed": 1.4,
            "glowIntensity": 1.8,
            "particleColor": "#F59E0B",
            "particleType": "cyber_embers"
        }
    }
]

def list_remotion_vfx(json_mode=False):
    if json_mode:
        return json.dumps({"compositions": REMOTION_COMPOSITIONS}, indent=2)

    output = "\n✨ VERCEL + REMOTION CLOUD VFX COMPOSITIONS (SERVERLESS GPU CLUSTER)\n========================================================================\n"
    for idx, comp in enumerate(REMOTION_COMPOSITIONS):
        output += f"{idx + 1}. [{comp['alias']} / {comp['id']}]\n"
        output += f"   Nama        : {comp['displayName']}\n"
        output += f"   Tipe Track  : {comp['targetTrackType']} Track | Durasi Default: {comp['defaultDurationMs'] / 1000}s\n"
        output += f"   Deskripsi   : {comp['description']}\n\n"
    output += "========================================================================\n"
    return output

def render_remotion_cloud_vfx(
    vfx_type="kinetic",
    title="FLOWMONKEY CLOUD VFX",
    subtitle="RENDERED WITH REMOTION ON VERCEL",
    theme_color="#8B5CF6",
    duration_sec=None,
    track_index=None,
    json_mode=False
):
    db = get_db()
    active_proj_id = db.get("activeProjectId") or (db["projects"][0]["id"] if db.get("projects") else 1)

    matched_comp = next(
        (c for c in REMOTION_COMPOSITIONS if c["id"].lower() == vfx_type.lower() or c["alias"].lower() == vfx_type.lower()),
        REMOTION_COMPOSITIONS[0]
    )

    duration_ms = int(float(duration_sec) * 1000) if duration_sec else matched_comp["defaultDurationMs"]

    # Find or create target track
    tracks = [t for t in db.get("tracks", []) if t.get("projectId") == active_proj_id and t.get("type") == matched_comp["targetTrackType"]]
    if tracks:
        track = tracks[0]
    else:
        new_track_id = max([t["id"] for t in db.get("tracks", [])], default=0) + 1
        track = {
            "id": new_track_id,
            "projectId": active_proj_id,
            "trackIndex": len([t for t in db.get("tracks", []) if t.get("projectId") == active_proj_id]),
            "type": matched_comp["targetTrackType"],
            "label": f"Remotion Cloud VFX ({matched_comp['targetTrackType']})"
        }
        db.setdefault("tracks", []).append(track)

    track_clips = [c for c in db.get("clips", []) if c.get("projectId") == active_proj_id and c.get("trackId") == track["id"]]
    start_time_ms = max([c.get("endTimeMs", 0) for c in track_clips], default=0)
    end_time_ms = start_time_ms + duration_ms

    job_id = f"remotion_{matched_comp['alias']}_{int(time.time() * 1000)}"
    cloud_cdn_url = f"https://storage.googleapis.com/flowmonkey-vfx-cloud/{matched_comp['id'].lower()}_{job_id}.mp4"

    new_clip_id = max([c["id"] for c in db.get("clips", [])], default=0) + 1
    new_clip = {
        "id": new_clip_id,
        "projectId": active_proj_id,
        "trackId": track["id"],
        "title": f"[Remotion Cloud] {matched_comp['id']}: {title[:20]}",
        "uri": cloud_cdn_url,
        "startTimeMs": start_time_ms,
        "endTimeMs": end_time_ms,
        "durationMs": duration_ms,
        "filterName": "Remotion Cloud HDR",
        "effectName": matched_comp["id"],
        "speedMultiplier": 1.0,
        "speedCurve": "Standard",
        "hasKeyframe": False,
        "keyframes": [],
        "transitionType": "None",
        "volume": 1.0,
        "isMuted": False,
        "noiseReduction": False,
        "vocalEnhance": False,
        "proxyUri": f"proxy_{job_id}.mp4",
        "proxyStatus": "READY",
        "cloudMetadata": {
            "renderer": "Vercel Serverless + Remotion GPU Cluster",
            "compositionId": matched_comp["id"],
            "inputProps": {
                "title": title,
                "subtitle": subtitle,
                "themeColor": theme_color
            },
            "renderedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        }
    }

    db.setdefault("clips", []).append(new_clip)
    save_db(db)

    if json_mode:
        return json.dumps({
            "success": True,
            "jobId": job_id,
            "compositionId": matched_comp["id"],
            "clip": new_clip,
            "streamUrl": cloud_cdn_url
        }, indent=2)

    return f"""
🚀 VERCEL + REMOTION CLOUD VFX SELESAI DI-RENDER!
========================================================================
Komposisi   : {matched_comp['displayName']} ({matched_comp['id']})
Job ID      : {job_id}
Judul/Teks  : "{title}"
Warna Tema  : {theme_color}
Durasi      : {(duration_ms / 1000):.1f}s (Track {track['type']} #{track['id']})
Timeline    : Ditambahkan pada rentang {(start_time_ms / 1000):.1f}s - {(end_time_ms / 1000):.1f}s
Cloud URL   : {cloud_cdn_url}
========================================================================
"""

def preview_remotion_cloud_vfx(vfx_type="kinetic", frame=30, json_mode=False):
    matched_comp = next(
        (c for c in REMOTION_COMPOSITIONS if c["id"].lower() == vfx_type.lower() or c["alias"].lower() == vfx_type.lower()),
        REMOTION_COMPOSITIONS[0]
    )
    preview_thumbnail = f"https://storage.googleapis.com/flowmonkey-vfx-cloud/previews/{matched_comp['id']}_f{frame}.jpg"

    if json_mode:
        return json.dumps({
            "success": True,
            "compositionId": matched_comp["id"],
            "previewFrame": int(frame),
            "previewThumbnail": preview_thumbnail,
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        }, indent=2)

    return f"🖼️ Keyframe Preview [Frame #{frame}] untuk '{matched_comp['id']}':\n   URL: {preview_thumbnail}"
